### ISSUE TITLE
Abandon Future handling and replace it with CompletionStage

### ISSUE THREAD

**Issue by rmannibucau** [eclipse/microprofile-fault-tolerance#362]
Today in java futures are most of the time CompletionStage (CompletionFutures), this is what the user can manipulate the most easily, this also enable any framework (MP or not, I'm thinking to DeltaSpike here) to integrate with the pipeline very easily.

Future on the other side does not enable it without reimplementing CompletionStage logic and creating some custom vendor specific API.

To rationalize this effort and promote java 8 I'd like fault-tolerance to replace its Future<?> usage by CompletionStage<?> which does not remove any feature from an user point of view?

**Comment by Emily-Jiang** [comment]
For backward compatibility issues, we are not going to remove Future but adding support for CompletionStage. I image most developers will move towards CompletionStage.

**Comment by rmannibucau** [comment]
@Emily-Jiang hmm, maybe I misunderstood but MP has no backward compatibility concern, metrics API was broken for instance (unit went from plain value to optional recently), is it really something taken into consideration?

**Comment by Emily-Jiang** [comment]
Good idea. We will add a javadoc or spec statement to emphasize the preference for CompletionStage.

**Comment by brunobat** [comment]
During the discussion of #368 the abandonment of Future looks inevitable due to implementation constraints.

**Comment by Azquelt** [comment]
I don't think #368 necessitates abandoning Future.

We can keep the existing behaviour for Future where the fault tolerance logic is applied around the method call, but we can't apply the fault tolerance logic around the completion of the Future, like we can for CompletionStage.

I don't see any reason why we should break backwards compatibility for a feature that people may already be using.

**Comment by brunobat** [comment]
Right, older methods using Future, will not take advantage of most Fault Tolerance functionalities. Something to document.

**Comment by Emily-Jiang** [comment]
Will add some sentences later with a PR to express the preference of CompletionStage.

**Comment by OndrejM** [comment]
Reviewing the spec, `Future` is basically handled only internally by the container and only when an app calls `Future.get()`, which has the same semantics as methods returning synchronously, except the app can choose when to apply those semantics. I think this is fine and covers `Future` enough, especially since most people would use `CompletionStage` instead of `Future` nowadays. Even `@Retry` can be supported like this, because once `Future.get()` is called, the thread is blocked and can wait for downstream futures and retry once they return. We just need to specify this more clearly or specify which annotations don't support the semantics if we decide to make implementers' life easier.

After we do this, I would conserve any support for `Future` and only focus on `CompletionStage`. 

I'm also not against removing support for `Future` altogether. I just don't like things in limbo, and currently a lot of semantics when using `Future` are underspecified and unclear.

So I would support one of the options:

* remove (deprecate) support for `Future` - I expect that with how its semantics are currently specified there are not many people using it with FT, and those, who are, struggle and would happily migrate to CS
* leave `Future` and specify semantics clearly, ideally that all functionality is supported and works like synchronous methods but is deferred to the call of `Future.get()`

Do you see any other option? Who thinks it's worth to invest in the second option?

**Comment by rmannibucau** [comment]
Works for me.

Side note: maybe the cancellation (timeout) should be clarified since it does not work like a synchronous method here and is up to the vendor currently (which is likely good cause only the behavior is important for end user, not the way it is done)

**Comment by Azquelt** [comment]
> that all functionality is supported and works like synchronous methods but is deferred to the call of Future.get()

I don't like this. If you're running something asynchronously, it's presumably because it takes some time, so it doesn't make sense to wait until the user asks for the result to realise that it didn't work and needs to be retried.

I think we should clearly specify that the functionality for Future is more limited than for completion stage and that the fault tolerance logic is only applied around the method execution, and not around the Future that is returned by the user's code. We should highlight that it's expected that the user's code will return a Future that is _already complete_.

This meets the following use cases:
* The user wants to run several slow blocking operations in parallel (e.g. using a synchronous HTTP client, or using a synchronous database driver)
* The user wants to use `@Timeout` with an operation which does not respond to the thread being interrupted

In contrast, CompletionStage also fulfils this use case:
* The user wants to apply fault tolerance logic around a call to an asynchronous API which returns a CompletionStage (e.g. JAX-RS 2.1 client using `.rx()`)

**Comment by Emily-Jiang** [comment]
Well said @Azquelt ! As I mentioned earlier, I'll just clarify in the spec to recommend CompletionStage for better async handling.
